Add default keyboard mappings for flattened object nav scripts

LeonarddeR opened the pull request:
### Link to issue number:
Fixes #15053 

### Summary of the issue:
Object navigation can be difficult to grasp for some people. It is helpful to assign the flattened object nav gestures (that are now only available on touch screen)by default.

### Description of user facing changes
See change log items

### Description of development approach
Added gestures to scripts in `globalCommands`.

### Testing strategy:
Test desktop and laptop assignments.

### Known issues with pull request:
Translators must be made aware of these new assignments, since they might need a special localized mapping for other keyboard layouts.

### Change log entries:
Changes
- Added default keyboard gesture assignments to move to the next or previous object in a flattened view of the object hierarchy. (#15053)
Add the following additional gesture mappings:
 - Desktop:NVDA+numpad9 and NVDA+numpad3 to move to the previous and next objects, respectively
 - Laptop: shift+NVDA+[ and shift+NVDA+] to move to the previous and next objects, respectively

### Code Review Checklist:

<!--
This checklist is a reminder of things commonly forgotten in a new PR.
Authors, please do a self-review of this pull-request.
Check items to confirm you have thought about the relevance of the item.
Where items are missing (eg unit / system tests), please explain in the PR.
To check an item `- [ ]` becomes `- [x]`, note spacing.
You can also check the checkboxes after the PR is created.
A detailed explanation of this checklist is available here:
https://github.com/nvaccess/nvda/blob/master/devDocs/githubPullRequestTemplateExplanationAndExamples.md#code-review-checklist
-->

- [x] Pull Request description:
  - description is up to date
  - change log entries
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] API is compatible with existing add-ons.
- [x] Documentation:
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] Security precautions taken.

LeonarddeR committed:
Add default keyboard mappings for flattened object nav scripts

cary-rowen commented:
Hi,

This set of keymaps has been used in the NVDA Chinese community for a long time.
For laptop layout we are using NVDA+Ctrl+PageUp/PageDown

leonardder commented:
I personally prefer to avoid adding pageUp/pageDown on laptop because many laptops don't have these keys.

zstanecic commented:
There is also one drawback. Page up and down are mapped on some laptops to fn+arrows.

That means, that it can be uncomfprtable for \users

 

From: Leonard de Ruijter ***@***.***> 
Sent: Wednesday, June 28, 2023 9:49 AM
To: nvaccess/nvda ***@***.***>
Cc: Subscribed ***@***.***>
Subject: Re: [nvaccess/nvda] Add default keyboard mappings for flattened object nav scripts (PR #15065)

 

I personally prefer to avoid adding pageUp/pageDown on laptop because many laptops don't have these keys.

—
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/15065#issuecomment-1610934208> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/ACVCDE2BSWYGEV4HADOLFXDXNPOWFANCNFSM6AAAAAAZWTUVUE> .
You are receiving this because you are subscribed to this thread.  <https://github.com/notifications/beacon/ACVCDEZKIX5PHTCJ3MDBW7LXNPOWFA5CNFSM6AAAAAAZWTUVUGWGG33NNVSW45C7OR4XAZNMJFZXG5LFINXW23LFNZ2KUY3PNVWWK3TUL5UWJTTAATT4A.gif> Message ID: ***@***.*** ***@***.***> >

leonardder commented:
I wonder whether we shouldn't yet go for the brackets for laptop instead. They are only one row above the current keys and might be a bit more obvious, i.e. clearly pointing left and right) and we don't have to do the accute/cedilla trick.

burmancomp commented:
> I wonder whether we shouldn't yet go for the brackets for laptop instead. They are only one row above the current keys and might be a bit more obvious, i.e. clearly pointing left and right) and we don't have to do the accute/cedilla trick.

Do you mean curly brackets ({}) which are shift+[ and shift+]? In Finnish key layout, key which writes } (shift+]) in US key layout, is somekind of key combination which is used to write special letters like û, ê and î. Letters are written so that first "}" key combination is pressed and then usual letter, for example  "}" key u gives û.

I'm not localizing Finnish version but I would like to know if this would be a problem if these keys would be tried to be added to localized gesture map. These keys should be unassigned currently in Finnish version.

Different keys seem to be assigned according to key layout of the language. For example keys producing ; and ' in US layout are with NVDA key assigned to mouse clicks in Finnish localization, and mouse clicks when NVDA language is English are (with NVDA key) keys producing [ and ] in US key layout.

CyrilleB79 commented:
> > I wonder whether we shouldn't yet go for the brackets for laptop instead. They are only one row above the current keys and might be a bit more obvious, i.e. clearly pointing left and right) and we don't have to do the accute/cedilla trick.
> 
> Do you mean curly brackets ({}) which are shift+[ and shift+]? In Finnish key layout, key which writes } (shift+]) in US key layout, is somekind of key combination which is used to write special letters like û, ê and î. Letters are written so that first "}" key combination is pressed and then usual letter, for example "}" key u gives û.
> 
> I'm not localizing Finnish version but I would like to know if this would be a problem if these keys would be tried to be added to localized gesture map. These keys should be unassigned currently in Finnish version.
> 
> Different keys seem to be assigned according to key layout of the language. For example keys producing ; and ' in US layout are with NVDA key assigned to mouse clicks in Finnish localization, and mouse clicks when NVDA language is English are (with NVDA key) keys producing [ and ] in US key layout.

We can see in the various `gestures.ini` files that gestures using keys other than letters (mainly punctuation) need to be localized according to locale keyboard layout. That's the case for laptop layout left or right mouse click.

Regarding dead key (i.e. key that do not type a character but change the next one), in French we have "^" located just at the right of "P" key, i.e. where "[" appears on US layout; this dead key is used followed by another vowel to type "â", "ê", "î", etc. I can see that it has not been used to localize for left click. @MichelSuch is there any reason why?

Of course, the best way to know if dead keys can be used is to test.

leonardder commented:
@burmancomp I meant the left and right brackets, i.e. [ and ].
As @CyrilleB79 pointed out, some magic happens in gestures.ini regarding to these assignments. The acute/cedilla should probably be be moved to the Dutch assignments and removed from the pr.

burmancomp commented:
> @burmancomp I meant the left and right brackets, i.e. [ and ].

Do I missunderstand something but are not [ and ] assigned to mouse clicks when nvda language is English?

leonardder commented:
> > @burmancomp I meant the left and right brackets, i.e. [ and ].
> 
> Do I missunderstand something but are not [ and ] assigned to mouse clicks when nvda language is English?

Yes, but not with shift added to the gesture 😉

burmancomp commented:
> > 
> > Do I missunderstand something but are not [ and ] assigned to mouse clicks when nvda language is English?
> 
> Yes, but not with shift added to the gesture 😉

And it means that we are speaking about curly brackets ({}) because they are produced from square brackets ([]) in uS layout andding shift.

CyrilleB79 commented:
> > > Do I missunderstand something but are not [ and ] assigned to mouse clicks when nvda language is English?
> > 
> > 
> > Yes, but not with shift added to the gesture 😉
> 
> And it means that we are speaking about curly brackets ({}) because they are produced from square brackets ([]) in uS layout andding shift.

Actually, when talking about key combination gestures in NVDA, we indicate the main character and the modifiers. So we usually do not write 'Type "{"', but 'Type "shift+["'. This goes for the User's Guide as well as for the code. It doesn't matter which character is ultimately produced.

LeonarddeR committed:
Switch to brackets

LeonarddeR committed:
Merge remote-tracking branch 'origin/master' into flattenedReview

seanbudd reviewed (approved):
Thanks @leonardder

Qchristensen reviewed (changes requested):
The text looks good, but two comments:

In "Section 5.6 Review Modes", it mentions "Review modes are a replacement for the older Flat Review concept found in NVDA.".  Given we are now introducing a "new" flat review and also as it has been 10 years (2013.2) since the original flat review was removed, could we also remove that sentence in this PR?

And in Section 5.4 Object Navigation (before the keystrokes) we have a paragraph which goes through the concept of the hierarchy of objects (beginning with "Rather than having to move back and forth through....") - is it worth a sentence at the end of that paragraph noting the new commands which essentially ignore it?

LeonarddeR committed:
Merge remote-tracking branch 'origin/master' into flattenedReview

LeonarddeR committed:
Docs review

leonardder commented:
@Qchristensen thanks for the valuable comments. I removed the said line and also added another paragraph about flat view object nav. Any suggestions to make this clearer are greatly appreciated.

seanbudd commented on user_docs/en/userGuide.t2t:
```diff
@@ -592,6 +592,12 @@ Moving to a list item's containing object will take you back to the list.
 You can then move past the list if you wish to access other objects.
 Similarly, a toolbar contains controls, so you must move inside the toolbar to access the controls in the toolbar.
 
+If you yet prefer to move back and forth between every single object on the system, you can use the move to previous/next object in a flattened view commands, respectively.
```
I think this will read slightly better

```suggestion
If you yet prefer to move back and forth between every single object on the system, you can use commands to move to the previous/next object in a flattened view.
```

Qchristensen reviewed (approved):
Thanks Leonard, great work!

LeonarddeR committed:
Update user_docs/en/userGuide.t2t

Co-authored-by: Sean Budd <seanbudd123@gmail.com>

seanbudd committed:
Merge remote-tracking branch 'origin/master' into flattenedReview

seanbudd committed:
update changes
